ag_data/formulas/ingestion_engine.py

```diff
@@ -0,0 +1,33 @@
+from ag_data import models
+from ag_data.formulas.library.system import mercury_formulas as hgFormulas
+
+
+class MeasurementIngestionEngine:
+
+    processingFormulas = {
+        0: hgFormulas.fPass,
+        2: hgFormulas.fMercurySimpleTemperatureSensor,
+        4: hgFormulas.fMercuryDualTemperatureSensor,
+        6: hgFormulas.fMercuryFlowSensor,
+    }
+
+    def saveMeasurement(self, measurementDict, event):
+        timestamp = measurementDict["measurement_timestamp"]
+        sensor_id = measurementDict["measurement_sensor"]
+        rawValue = measurementDict["measurement_values"]
+
+        sensor = models.AGSensor.objects.get(pk=sensor_id)
+        sensor_type = sensor.type_id
+        processing_formula = sensor_type.processing_formula
```

Comment:
To make sure that I understand correctly, this `sensor_type.processing_formula` is expected to be one of `0, 2, 4, 6`?
If so, how is this value guaranteed to be one of them?

~My guessing is that our server somehow sends a user a list of available numbers (=0, 2, 4, 6) and there's no way for a user to choose a value other than the one in the list. But I couldn't find any link from the key of `processingFomula` dictionary (0, 2, 4, 6).~

~Could you explain this?~

Reply:
I found that there is `built_in_content.py` and also `ag_data/presets/sample_user_data.py`. I guess the implementation of the builtin function and also `processingFormula` dictionary should be inside the file.
And there are total 3 files that hardcode 0, 2, 4, 6. It makes maintenance harder. For instance, when you add a sensor with index 10, then [you should edit three files (Shotgun surgery)](https://en.wikipedia.org/wiki/Shotgun_surgery) and it's not even easy to find which files to edit.

Reply:
This is a fast and cheap but not good implementation. Ideally, the mapping should be robust and easy to change, so the user or developer does not need to maintain the digits in different yet not directly linked files to modify the code and still make it work. I am thinking what Dan has been saying, that a string or other version of mapping may be more appropriate. 

The problem here is, how do we map something defined in a dictionary (maybe a value of some dictionary-allowed type, such as numbers, strings, bool, etc) to a function handler?

Reply:
It's obvious that we should have this kind of in-memory dictionary that maps something to a function in our python code unless we dynamically evaluate a raw string from outside of our codebase hoping that it's a valid python code snippets. 

I don't really want to evaluate an untrusted code. Providing a sandboxed environment for it is tricky especially for python.

So the conclusion is, the way that the current code works is good as-is but let's organize it for us not to edit multiple locations when we work on it.

Reply:
I feel this topic is bigger than what we can accommodate in the discussion of a PR. So here is a dedicated issue #378 created for reimagining the formula function mapping.